### PR TITLE
Core Type #2 - Core changes

### PR DESCRIPTION
**Pull request by justinkaseman**

### CODE REVIEW

**Review by cds95 — APPROVED** [review]

File: packages/core/bootstrap/src/lib/config/index.ts [hunk 1]
```diff
@@ -20,10 +20,17 @@ export const constants = {
 }
 
 /* eslint-disable @typescript-eslint/no-unused-vars */
-const cloneNoSecrets = (config: Config): Config =>
-  (({ apiKey, api: { auth, headers, params, ...api }, ...o }) => ({ api, ...o }))(config)
+const cloneNoSecrets = (config: Partial<Config>): Partial<Config> =>
+  (({ apiKey, api, ...o }) => {
+    const { auth, headers, params, ...apiNoSecrets } = api || {}
+    return { api: apiNoSecrets, ...o }
```
Comment (boxhock) [inline packages/core/bootstrap/src/lib/config/index.ts:26]: Appreciate the added readability here!

File: packages/core/bootstrap/src/lib/middleware/cache-warmer/epics.ts [hunk 2]
```diff
@@ -285,19 +288,19 @@ export const warmupRequestHandler: Epic<AnyAction, AnyAction, any> = (action$, s
...
-    filter(
-      ([{ payload }, state]) =>
-        (state.cacheWarmer.warmups[payload.key]?.errorCount ?? 0 >= config.unhealthyThreshold) &&
-        config.unhealthyThreshold !== -1,
-    ),
+    filter(([{ payload }, state]) => {
+      if (config.unhealthyThreshold < 0) return false
+      const errorCount = state.cacheWarmer.warmups[payload.key]?.errorCount ?? 0
+      return errorCount >= config.unhealthyThreshold
+    }),
```
Comment (boxhock) [inline packages/core/bootstrap/src/lib/middleware/cache-warmer/epics.ts:303]: Again, so much easier to read :pray:

File: packages/core/bootstrap/src/lib/middleware/cache-warmer/index.ts [hunk 3]
```diff
@@ -8,45 +14,52 @@ import { getWSConfig } from '../ws/config'
...
+  ) =>
+  (rawExecute: Execute<AdapterRequest<D>>): Middleware<AdapterRequest<D>> =>
   async (execute, context) =>
-  async (input: AdapterRequest) => {
+  async (input) => {
     const isWarmerActive =
-      util.parseBool(util.getEnv('CACHE_ENABLED') ?? DEFAULT_CACHE_ENABLED) &&
-      util.parseBool(util.getEnv('WARMUP_ENABLED') ?? DEFAULT_WARMUP_ENABLED)
+      util.parseBool(util.getEnv('CACHE_ENABLED', undefined, context)) &&
+      util.parseBool(util.getEnv('WARMUP_ENABLED'))
```
Comment (boxhock) [inline packages/core/bootstrap/src/lib/middleware/cache-warmer/index.ts:44]: Same here regarding defaults

File: packages/core/bootstrap/test/unit/cache.test.ts [hunk 4]
```diff
@@ -105,36 +110,175 @@ describe('cache', () => {
...
-      await callAndExpect(counter, 1000, 1)
+    //   clock.tick(1000 * 5 + 1) // extra 1ms
+    //   await callAndExpect(counter, 1, 0)
+    //   await callAndExpect(counter, 1000, 0)
 
-      clock.tick(1000 * 35 + 1) // extra 1ms
-      await callAndExpect(counter, 1, 2)
+    //   clock.tick(1000 * 30 + 1) // extra 1ms
+    //   await callAndExpect(counter, 1, 1)
+    // })
```
Comment (boxhock) [inline packages/core/bootstrap/test/unit/cache.test.ts:143]: Why is this commented out?

**Review by boxhock — COMMENTED** [review]
Looks good! There are small but very meaningful improvements all over the place.
The PR is still a bit massive (+6,335 −1,696), but I'm guessing we can't reduce this further, so will likely require some more eyes on this.

**Comment by austinborn** [comment]
@justinkaseman Can you rebase and update this? Don't want it to sit too long...

File: packages/core/bootstrap/test/unit/cache.test.ts (hunk 4 above)
Comment (justinkaseman) [reply packages/core/bootstrap/test/unit/cache.test.ts]: Shouldn't have been. Un-commented it out.

File: packages/core/bootstrap/src/lib/middleware/normalize/index.ts [hunk 5]
```diff
@@ -1,25 +1,67 @@
...
+            description: 'The symbol of the currency to query',
+
+            required: true,
+
+            type: 'string',
+        },
+
+    }
+
+    Incoming `from` or `coin` keys would be renamed to `base`.
```
Comment (boxhock) [inline packages/core/bootstrap/src/lib/middleware/normalize/index.ts:32]: Seems like we have double newlines here, intentional? :sweat_smile:

File: packages/core/bootstrap/src/lib/metrics/util.ts [hunk 6]
```diff
@@ -80,8 +80,20 @@ export const getFeedId = (input: AdapterRequest): string => {
...
+            (base as Array<unknown>).every((entry) => typeof entry === 'string'))) &&
+        (typeof quote === 'undefined' ||
+          typeof quote === 'string' ||
+          (Array.isArray(quote) &&
+            (quote as Array<unknown>).every((entry) => typeof entry === 'string')))
+      ) {
+        return (
+          `${buildSymbolString(base as string | string[])}` +
+          (quote ? `/${buildSymbolString(quote as string | string[])}` : '')
+        )
```
[the remark this answers is not in the document]
Comment (boxhock) [reply packages/core/bootstrap/src/lib/metrics/util.ts:96]: Could write a helper function

File: packages/core/bootstrap/src/lib/middleware/normalize/index.ts (hunk 5 above)
Comment (justinkaseman) [reply packages/core/bootstrap/src/lib/middleware/normalize/index.ts:32]: JSDoc limitation :disappointed:

**Pull request merged**